README.md
# Lite Switch
Lite Switch is a customized switch which you can add or change text and icon for "On" and "Off".
You can also set the width and height for the switch in your app.
Check github: https://github.com/galacticpuffin/lite_switch

## Features
* [New] Trigger the switch to change through the ValueNotifier<bool> SwitchState. So you can control how the switch shows from your app.
* Get the switch value change through implementing the onChanged.
* Set iWidth and iHeight to get the right size for your switch.
* Customize your text and icon for switch "On" and "Off".

## Screenshot
 <img height="414" src="https://github.com/galacticpuffin/lite_switch/raw/master/assets/images/on.png">
 <img height="414" src="https://github.com/galacticpuffin/lite_switch/raw/master/assets/images/off.png">
 <img height="414" src="https://github.com/galacticpuffin/lite_switch/raw/master/assets/images/switchStateOff.png">
 <img height="414" src="https://github.com/galacticpuffin/lite_switch/raw/master/assets/images/switchStateOn.png">

## Usage
To use this plugin :
```dart
Add the dependency to your pubspec.yaml file.
  dependencies:
    flutter:
      sdk: flutter
      lite_switch:
```

## Example
```dart
import 'package:flutter/material.dart';
import 'package:lite_switch/lite_switch.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _LiteSwitchState createState() => new _LiteSwitchState();
}

class _LiteSwitchState extends State<MyApp> {
  final ValueNotifier<bool> SwitchState = new ValueNotifier<bool>(true);
  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      body: new Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            LiteSwitch(
              switchState: SwitchState,
              initValue: true,
              textSize: 14.0,
              iWidth: 140,
              iHeight: 60,
              textOn: 'Sound ON',
              textOff: 'Sound Off',
              colorOn: Colors.deepOrange,
              colorOff: Colors.black,
              iconOn: Icons.volume_up,
              iconOff: Icons.volume_off,
              onChanged: (bool state) {
                setState(() {
                  SwitchState.value = !SwitchState.value;
                  print('turned ${(state) ? 'on' : 'off'}');
                });
              },
            ),
            RaisedButton(
              onPressed: () {
                setState(() {
                  SwitchState.value = !SwitchState.value;
                  print(
                      'onPressed ${(SwitchState.value) ? 'on' : 'off'}');
                });
              },
              child: Text("On/Off"),
            ),
            Text("State:${(SwitchState.value) ? 'on' : 'off'}"),
          ],
        ),
      ),
    );
  }
}
```

CHANGELOG.md
## [1.0.6] Apri 16, 2020
* Add new ValueNotifier<bool> switchState to listen to the switch state changes. So when parent widge change the switchState to true, lite_switch will animated to "On".
when parent widget changet switchState to false, lite_switch will animated to "Off"
It can use switchState if you app need want to control the switch.

## [1.0.5] Apri 14, 2020
* Fix the value change event bug.

## [1.0.4] Apri 14, 2020
* Change class name using UpperCamelName
* Update Package description
* fix screenshot link bug
* Update License
* Add Example

## [1.0.3] Apri 14, 2020

* Fix screenshot link bug

## [1.0.2] Apri 14, 2020

* Fix screenshot link

## [1.0.1] Apri 14, 2020

* Add screenshot

## [1.0.0] Apri 13, 2020

* Init the Customized Switch


test/lite_switch_test.dart
import 'package:flutter/material.dart';
import 'package:lite_switch/lite_switch.dart';

void main() => runApp(testswitch());
class testswitch extends StatefulWidget {
  @override
  _LiteSwitchState createState() => new _LiteSwitchState();
}

class _LiteSwitchState extends State<testswitch> {

  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      body: new Center(
        child: new Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Padding(
              padding: EdgeInsets.only(top: 0),
              child: LiteSwitch(
                initValue: true,
                textSize: 16.0,
                iWidth: 140,
                iHeight: 40,
                textOn: 'Sound ON',
                textOff: 'Sound Off',
                colorOn: Colors.deepOrange,
                colorOff: Colors.black,
                iconOn: Icons.volume_up,
                iconOff: Icons.volume_off,
                onChanged: (bool state) {
                },
              ),
            ),
          ],
        ),
      ),
    );
  }
}



lib/lite_switch.dart
import 'package:flutter/material.dart';
import 'dart:ui';

class LiteSwitch extends StatefulWidget {
  @required
  final bool initValue;
  @required
  final Function(bool) onChanged;

  final double iWidth;
  final double iHeight;
  final String textOff;
  final String textOn;
  final Color colorOn;
  final Color colorOff;
  final double textSize;
  final Duration animationDuration;
  final IconData iconOn;
  final IconData iconOff;
  final Function onTap;
  final Function onDoubleTap;
  final Function onSwipe;
  // Change SwitchState will notify the switch animation to ON or Off if needed.
  final ValueNotifier<bool> switchState;

  LiteSwitch(
      {this.switchState,
      this.initValue = true,
      this.iWidth = 120,
      this.iHeight = 40,
      this.textOff = "Off",
      this.textOn = "On",
      this.textSize = 16.0,
      this.colorOn = Colors.deepOrange,
      this.colorOff = Colors.black,
      this.iconOff = Icons.volume_off,
      this.iconOn = Icons.volume_up,
      this.animationDuration = const Duration(milliseconds: 100),
      this.onTap,
      this.onDoubleTap,
      this.onSwipe,
      this.onChanged});

  @override
  _LiteSwitchState createState() => _LiteSwitchState();
}

class _LiteSwitchState extends State<LiteSwitch>
    with SingleTickerProviderStateMixin {
  AnimationController animationController;
  Animation<double> animation;
  double value = 0.0;
  double get offsetvalue {
    if (value < 0.5) return -1;
    return 1;
  }

  bool turnState;

  @override
  void dispose() {
    animationController.dispose();
    super.dispose();
  }

  @override
  void initState() {
    super.initState();

    animationController = AnimationController(
        vsync: this,
        lowerBound: 0.0,
        upperBound: 1.0,
        duration: widget.animationDuration);
    animation =
        CurvedAnimation(parent: animationController, curve: Curves.easeInOut);
    animationController.addListener(() {
      setState(() {
        //print("animationController.addListener");
        value = animation.value;
      });
    });
    turnState = widget.initValue;
    _determine();
  }

  ValueNotifier valueNotifier;
  @override
  Widget build(BuildContext context) {
    Color transitionColor = Color.lerp(widget.colorOff, widget.colorOn, value);
    widget.switchState.addListener(() {
      setState(() {
        //print('trigger');
        bool newValue = widget.switchState.value;
        (newValue)
            ? animationController.forward()
            : animationController.reverse();
      });
    });
    return GestureDetector(
      onDoubleTap: () {
        _action();
        if (widget.onDoubleTap != null) widget.onDoubleTap();
      },
      onTap: () {
        _action();
        if (widget.onTap != null) widget.onTap();
      },
      onPanEnd: (details) {
        _action();
        if (widget.onSwipe != null) widget.onSwipe();
      },
      child: Container(
        padding: EdgeInsets.all(5),
        width: widget.iWidth,
        height: widget.iHeight,
        decoration: BoxDecoration(
            color: transitionColor, borderRadius: BorderRadius.circular(40)),
        child: Stack(
          children: <Widget>[
            Transform.translate(
              offset: Offset(10 * value, 0),
              child: Opacity(
                opacity: (1 - value).clamp(0.0, 1.0).toDouble(),
                child: Container(
                  padding: EdgeInsets.only(right: 10),
                  alignment: Alignment.centerRight,
                  child: Text(
                    widget.textOff,
                    style: TextStyle(
                        color: Colors.white,
                        fontWeight: FontWeight.bold,
                        fontSize: widget.textSize),
                  ),
                ),
              ),
            ),
            Transform.translate(
              offset: Offset(10 * (1 - value), 0),
              child: Opacity(
                opacity: value.clamp(0.0, 1.0).toDouble(),
                child: Container(
                  padding: EdgeInsets.only(left: 5),
                  alignment: Alignment.centerLeft,
                  child: Text(
                    widget.textOn,
                    style: TextStyle(
                        color: Colors.white,
                        fontWeight: FontWeight.bold,
                        fontSize: widget.textSize),
                  ),
                ),
              ),
            ),
            Transform.translate(
              offset: Offset(
                  (widget.iWidth / 2 - widget.iHeight / 2) * offsetvalue, 0),
              child: Container(
                alignment: Alignment.center,
                decoration:
                    BoxDecoration(shape: BoxShape.circle, color: Colors.white),
                child: Stack(
                  children: <Widget>[
                    Center(
                      child: Opacity(
                        opacity: (1.0 - value).clamp(0.0, 1.0).toDouble(),
                        child: Icon(
                          widget.iconOff,
                          size: widget.iHeight / 2,
                          color: transitionColor,
                        ),
                      ),
                    ),
                    Center(
                        child: Opacity(
                            opacity: value.clamp(0.0, 1.0).toDouble(),
                            child: Icon(
                              widget.iconOn,
                              size: widget.iHeight / 2,
                              color: transitionColor,
                            ))),
                  ],
                ),
              ),
              //),
            )
          ],
        ),
      ),
    );
  }

  void _action() {
    _determine(changeState: true);
  }

  void _determine({bool changeState = false}) {
    setState(() {
      if (changeState) turnState = !turnState;
      (turnState)
          ? animationController.forward()
          : animationController.reverse();
      widget.onChanged(turnState);
    });
  }
}


example/example.dart
import 'package:flutter/material.dart';
import 'package:lite_switch/lite_switch.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _LiteSwitchState createState() => new _LiteSwitchState();
}

class _LiteSwitchState extends State<MyApp> {
  final ValueNotifier<bool> SwitchState = new ValueNotifier<bool>(true);
  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      body: new Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            LiteSwitch(
              switchState: SwitchState,
              initValue: true,
              textSize: 14.0,
              iWidth: 140,
              iHeight: 60,
              textOn: 'Sound ON',
              textOff: 'Sound Off',
              colorOn: Colors.deepOrange,
              colorOff: Colors.black,
              iconOn: Icons.volume_up,
              iconOff: Icons.volume_off,
              onChanged: (bool state) {
                setState(() {
                  SwitchState.value = !SwitchState.value;
                  print('turned ${(state) ? 'on' : 'off'}');
                });

              },
            ),
            RaisedButton(
              onPressed: () {
                setState(() {
                  SwitchState.value = !SwitchState.value;
                  print(
                      'onPressed ${(SwitchState.value) ? 'on' : 'off'}');
                });
              },
              child: Text("On/Off"),
            ),
            Text("State:${(SwitchState.value) ? 'on' : 'off'}"),
          ],
        ),
      ),
    );
  }
}
